### PR TITLE
GH-123040: Specialize shadowed `LOAD_ATTR`.

### PR DESCRIPTION
**Pull request by markshannon**
This PR specializes `LOAD_ATTR` for shadowed attributes, where the shadowing doesn't change the attribute.

[Stats](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240821-3.14.0a0-975089f/bm-20240821-azure-x86_64-faster%252dcpython-specialize_shadowed_-3.14.0a0-975089f-pystats-vs-base.md) show a 30% reduction in specialization failures for `LOAD_ATTR`.

[Performance](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20240821-3.14.0a0-975089f) might show a slight speedup, but is more likely just noise.

* Issue: gh-123040

### CODE REVIEW

File: Python/specialize.c [hunk 1]
```diff
@@ -1009,73 +1024,112 @@ _Py_Specialize_LoadAttr(_PyStackRef owner_st, _Py_CODEUNIT *instr, PyObject *nam
...
+    if (!_PyType_IsReady(type)) {
+        // We *might* not really need this check, but we inherited it from
+        // PyObject_GenericGetAttr and friends... and this way we still do the
+        // right thing if someone forgets to call PyType_Ready(type):
+        SPECIALIZATION_FAIL(LOAD_ATTR, SPEC_FAIL_OTHER);
+        goto fail;
+    }
+    if (PyModule_CheckExact(owner)) {
+        if (specialize_module_load_attr(owner, instr, name))
+        {
```
Comment (iritkatriel) [inline Python/specialize.c]: You're mixing { at end of line sometimes and on its own line other times.

File: Python/specialize.c [hunk 2]
```diff
@@ -895,57 +902,65 @@ static int specialize_attr_loadclassattr(PyObject* owner, _Py_CODEUNIT* instr, P
...
     }
-    if (PyModule_CheckExact(owner)) {
-        if (specialize_module_load_attr(owner, instr, name))
-        {
-            goto fail;
-        }
-        goto success;
+    if (cls->tp_flags & Py_TPFLAGS_INLINE_VALUES) {
+        PyDictKeysObject *keys = ((PyHeapTypeObject *)cls)->ht_cached_keys;
+        int index = _PyDictKeys_StringLookup(keys, name);
```
Comment (iritkatriel) [inline Python/specialize.c]: compiler warning here

File: Python/specialize.c [hunk 3]
```diff
@@ -895,57 +902,65 @@ static int specialize_attr_loadclassattr(PyObject* owner, _Py_CODEUNIT* instr, P
     PyObject* descr, DescriptorClassification kind, bool is_method);
 static int specialize_class_load_attr(PyObject* owner, _Py_CODEUNIT* instr, PyObject* name);
 
-void
-_Py_Specialize_LoadAttr(_PyStackRef owner_st, _Py_CODEUNIT *instr, PyObject *name)
+/* Returns true if instance may have attribute `name`. */
```
Comment (iritkatriel) [inline Python/specialize.c]: may in the sense of maybe, or in the sense of can?

**Review by iritkatriel — COMMENTED** [review]
>[Stats](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240821-3.14.0a0-975089f/bm-20240821-azure-x86_64-faster%252dcpython-specialize_shadowed_-3.14.0a0-975089f-pystats-vs-base.md) show a 30% reduction in specialization failures for LOAD_ATTR.

> [Performance](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20240821-3.14.0a0-975089f) might show a slight speedup, but is more likely just noise.

Why is it useful to improve specialisation stats if it's not also improving performance?

File: Python/specialize.c (hunk 3 above)
Comment (markshannon) [reply Python/specialize.c]: Neither of those really. I'll clarify the comment.

**Comment by markshannon** [comment]
> Why is it useful to improve specialisation stats if it's not also improving performance?

It probably does improve performance, and we would expect it to, but it's too small to measure.
That's why we rely on the stats, as they are much less noisy.

We also expect the difference to be magnified on tier 2 once that is more complete.

File: Python/specialize.c (hunk 1 above)
Comment (markshannon) [reply Python/specialize.c]: I reorganized this to not use labels and gotos.

**Comment by iritkatriel** [comment]
LGTM, but I think another reviewer might make sense, @brandtbucher or @Fidget-Spinner maybe.

**Review by brandtbucher — APPROVED** [review]
This is some pretty subtle stuff, but it looks correct to me! I wasn't able to make it do anything incorrect with a local build.

**Pull request merged**